VXLAN: T4120: add ability to set multiple remotes

justsecure opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->
VXLAN does support using multiple remotes but VyOS does not.
Add the ability to set multiple remotes and add them using "bridge"
commands

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- All submitted PRs must be linked to a Task on Phabricator. -->
* https://phabricator.vyos.net/T4120

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->
vxlan

## Proposed changes
<!--- Describe your changes in detail -->
Instead adding vxlan interface like 
```
ip link add vxlan10 type vxlan id 10 remote 10.0.0.1 dstport 0
```

Insert them without remote and then add the remotes with "bridge" command:
```
ip link add vxlan10 type vxlan id 10 dstport 0
bridge fdb append to 00:00:00:00:00:00 dst 10.0.0.1 dev vxlan10
bridge fdb append to 00:00:00:00:00:00 dst 10.0.0.2 dev vxlan10
```
This allows using multiple remotes.
Broadcast traffic (like arp) got pushed to all remotes then. After a mac address is learned vxlan only sends to that remote.

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
-->
```
set interfaces vxlan vxlan10 remote '10.0.0.1'
set interfaces vxlan vxlan10 remote '10.0.0.2'
set interfaces vxlan vxlan10 vni '10'
commit

vyos@vyos# bridge fdb show dev vxlan10
00:00:00:00:00:00 dst 10.0.0.1 self permanent
00:00:00:00:00:00 dst 10.0.0.2 self permanent
```
If another port is used then it also works. You do not see the changed port with "bridge dfb show" but tcpdump prove that the other port is used.

Removing remotes is also possible:
```
delete interfaces vxlan vxlan10 remote '10.0.0.2'
commit

vyos@vyos# bridge fdb show dev vxlan10
00:00:00:00:00:00 dst 10.0.0.1 self permanent
```

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

Commit:
VXLAN: T4120: add ability to set multiple remotes
VXLAN does support using multiple remotes but VyOS does not.
Add the ability to set multiple remotes and add them using "bridge"
commands

dmbaturin reviewed (approved):
Looks sensible at a glance, but I haven't tested it myself yet.

sever-sever commented:
Is it overlapped with this commit? https://github.com/vyos/vyos-1x/pull/1126/commits/790cc11d405f3777e498d58d53a16548dbd9318f

justsecure commented:
> Is it overlapped with this commit? [790cc11](https://github.com/vyos/vyos-1x/commit/790cc11d405f3777e498d58d53a16548dbd9318f)

_No. They are completely independent. 
I hope I did not linked them.
Just learned today how to use git for pull requests and still learning ;)_

EDIT: pardon me. Now I see what was wrong. 
I have to use another branch for every PR.

sever-sever commented on interface-definitions/include/interface/tunnel-remotes.xml.i:
```diff
@@ -0,0 +1,19 @@
+<!-- include start from rip/tunnel-remote.xml.i -->
```
```
 <!-- include start from rip/tunnel-remote.xml.i -->
```
Replace to the correct include `from`

justsecure replied:
Sorry. I missed that. I hope it's okay now.

sever-sever replied:
It still has an incorrect include path.

justsecure replied:
Oh. That was already wrong on the "tunnel-remote.xml.i" I copied from.
I immediately correct that.

Am I allowed to correct both in one patch or do I have to create a different PR for correcting the other file?
I suppose the 2nd.

Commit:
VXLAN: T4120: add ability to set multiple remotes (rev2)
Replaced to the correct include from

Commit:
VXLAN: T4120: add ability to set multiple remotes (rev3)
Replaced to the correct path of include from

justsecure commented:
Found [two bugs](https://phabricator.vyos.net/T4120#115127) and working on them.
I hope I can supply a patch soon.

Commit:
VXLAN: T4120: add ability to set multiple remotes (rev4)

Fixed bug with multiple IPv6 remotes.
Now the first remote is left on dictionary and added by ip link command.
Other remotes (if existing) are moved to "other_remotes"
and added by bridge command

justsecure commented:
Major bug fixed. 
Now the first "remote" is left on dictionary and added by "ip link" command.
The other remotes (if more than one are defined) are moved to "other_remotes" and added by "bridge" command.
With that change it's possible to add multiple IPv6 remotes.

_There is still one bug left when IPv4 and IPv6 remotes are set on the same vxlan-interface.
I have to admit that I currently don't know how to prohibit that._
EDIT: done.

I also have to learn how to write (and execute) a smoketest and how to update the documentation.

Commit:
VXLAN: T4120: add ability to set multiple remotes (rev5) added checks

Added checks for mixed IPv4/IPv6 addresses on remotes and
source-interface

Commit:
VXLAN: T4120: adding smoketests with multiple IPv4/IPv6 remotes

c-po commented:
Implemented via 0ecddff7cffa - thanks
